### PR TITLE
Add nginx service

### PR DESCRIPTION
**Pull request by belka-ew**
It is an SBo package.

### CODE REVIEW

**Comment by mumahendras3** [comment]
Hi,

Sorry for the late response, and thank you for the PR!

I'll try to look at this when I have some spare time in the weekend.

Cheers!

**Comment by belka-ew** [comment]
I finally got some time to adapt the services I've written in the last years to your s6 templates and publish them, so they can be useful for other people. So I hope you don't mind if I create a few more pull requests in the next days.

**Comment by mumahendras3** [comment]
Heh, I don't mind at all. I'm glad if someone finds this repo useful.

Feel free submitting pull requests here, although I hope you don't mind my slow responses :D

**Comment by belka-ew** [comment]
Your responses aren't slow, I'm usually much slower to respond.

And yes, the repository is useful  and thank you for your work on it. I'm using s6 on my desktop and 2 Slackware servers even if the servers just have a few customers, they run a lot of smaller services.

**Comment by mumahendras3** [comment]
> And yes, the repository is useful and thank you for your work on it. I'm using s6 on my desktop and 2 Slackware servers even if the servers just have a few customers, they run a lot of smaller services.

Heh, thanks for the encouragement.

Anyway, finally got some time to look into this. Everything looks good to me, although it seems using `-g daemon off` makes the nginx master process (and, consequently, its worker processes) not redirect `stdout` to `/dev/null`. With the default nginx config, does it still use `stdout` for logging something? Have you ever seen nginx logs show up in the catch-all logger when using it in production? With my limited testing, it doesn't seem to use `stdout` for anything, so I guess we don't need a logger service for it.

**Comment by belka-ew** [comment]
Not often checking the catch-all logger, but no, nginx shouldn't write anything into it. Nginx has its own "catch-all" logger. Nginx is similar to Apache, it supports virtual hosts, so basically each website it serves can have its own logfile, or in my case I setup a log file for each customer. So its pretty impractical to use a single log file for nginx.

Basically configuring nginx you define a default log file:

```
	access_log /var/log/nginx/access.log main;
```

And then you can override this setting for specific hostnames:

```
server {
	listen …:443 ssl http2;

	server_name git.caraus.tech;
	root /srv/httpd/server/git;

	location / { 
            …
	}

	error_log /var/log/nginx/server.error error;
	access_log /var/log/nginx/server.access;
}
```

**Comment by belka-ew** [comment]
Just checked the default configuration. It uses `/var/log/nginx/error.log` and `/var/log/nginx/access.log` by default. But it will report critical errors to stderr like broken configuration or if it cannot bind to an address and port.

**Comment by mumahendras3** [comment]
> Just checked the default configuration. It uses `/var/log/nginx/error.log` and `/var/log/nginx/access.log` by default. But it will report critical errors to stderr like broken configuration or if it cannot bind to an address and port.

I see. I guess we should add a service logger for it, just in case a critical error occurs.

Well, actually, I have thought about this quite a while back, where every longrun service should have a logger service, since errors can occur even before `exec`-ing into the main binary (e.g., missing binaries). But, we'll worry about that later :D. For now, would you mind creating a logger service for this service (i.e., the corresponding `*-log` service)?

**Comment by belka-ew** [comment]
Sure, done.

**Review by mumahendras3 — CHANGES_REQUESTED** [review]
Awesome.

One more thing: you also need to create `nginx-srv/producer-for` that contains `nginx-log` to pair up with `nginx-log/consumer-for`, else `s6-rc-compile` will complain about it.

File: extras/rc.nginx/nginx-log/env/LOG_SCRIPT [hunk 1]
```diff
@@ -0,0 +1,2 @@
+n4 s500000 $LOGDIR_PATH
```
Comment (mumahendras3) [inline extras/rc.nginx/nginx-log/env/LOG_SCRIPT]: After testing this PR on my machine with default settings, it seems `nginx` doesn't add time stamps to logs printed on `stdout`. Here is an example output of `/var/log/nginx/current`:

```
nginx: [emerg] bind() to 0.0.0.0:22 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:22 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:22 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:22 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:22 failed (98: Address already in use)
nginx: [emerg] still could not bind()
```

That means, we have to add the `T` directive to our `LOG_SCRIPT`:

```suggestion
n4 s500000 T $LOGDIR_PATH
```

**Review by mumahendras3 — CHANGES_REQUESTED** [review]
Sorry, I missed this one small detail. Hopefully this is the last one :D.

File: extras/rc.nginx/nginx-log/env/LOG_SCRIPT (hunk 1 above)
Comment (belka-ew) [reply extras/rc.nginx/nginx-log/env/LOG_SCRIPT]: Right, haven't noticed that, thanks.

**Review by mumahendras3 — APPROVED** [review]

**Review by mumahendras3 — APPROVED** [review]
Okay, LGTM. Merging now.

Thanks again for the PR!

**Pull request merged**